gateway/proxy.go

```diff
@@ -68,21 +68,36 @@ func customDirector(targetURL *url.URL, originalDirector func(*http.Request)) fu
 	}
 }
 
-func customTransport(component string, timeouts Upstream) *http.Transport {
-	dialerTimeout := timeouts.HTTPClientDialerTimeout * time.Second
+func customTransport(component string, upstream Upstream) http.RoundTripper {
+	dialerTimeout := upstream.HTTPClientDialerTimeout * time.Second
 	if dialerTimeout == 0 {
 		dialerTimeout = defaultTimeoutValues[component].HTTPClientDialerTimeout
 	}
-	TLSHandshakeTimeout := timeouts.HTTPClientTLSHandshakeTimeout * time.Second
+	TLSHandshakeTimeout := upstream.HTTPClientTLSHandshakeTimeout * time.Second
 	if TLSHandshakeTimeout == 0 {
 		TLSHandshakeTimeout = defaultTimeoutValues[component].HTTPClientTLSHandshakeTimeout
 	}
-	responseHeaderTimeout := timeouts.HTTPClientResponseHeaderTimeout * time.Second
+	responseHeaderTimeout := upstream.HTTPClientResponseHeaderTimeout * time.Second
 	if responseHeaderTimeout == 0 {
 		responseHeaderTimeout = defaultTimeoutValues[component].HTTPClientResponseHeaderTimeout
 	}
+	dnsRefreshInterval := upstream.DNSRefreshInterval * time.Second
+	if dnsRefreshInterval == 0 {
+		dnsRefreshInterval = 1 * time.Second
+	}
+
+	url, err := url.Parse(upstream.URL)
+	if err != nil {
+		// TODO: log the error with logrus
+		fmt.Println(err)
+	}
+
+	t := &CustomTransport{
+		Transport: *http.DefaultTransport.(*http.Transport).Clone(),
+		lb:        newRoundRobinLoadBalancer(url.Hostname()),
+	}
+	go t.lb.refreshIPs(upstream.DNSRefreshInterval)
```

Comment:
this is not ideal, as there can be a panic inside DNSRefresh and we won't even know there was a panic.
But I am ok if you improve this in a follow up PR.

Reply:
I could not see how this code can produce a panic. Can you elaborate a little bit?

Reply:
Maybe not right now, but in general not waiting for a goroutine to return or panic will create trouble later as we add more code.

[See](https://github.com/cortexproject/auth-gateway/blob/main/server/server.go#L220) for example how server.go does it 

I find these resources valuable. It's tricky to get it right.

https://github.com/golang/go/wiki/LearnConcurrency
https://news.ycombinator.com/item?id=26509986

Reply:
I appreciate it for providing the resources. I will take a deep look into it